---
output:
  rmarkdown::github_document
bibliography: "inst/REFERENCES.bib"
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# R/`methyvim`

[![Travis-CI Build Status](https://travis-ci.org/nhejazi/methyvim.svg?branch=master)](https://travis-ci.org/nhejazi/methyvim)
[![AppVeyor Build  Status](https://ci.appveyor.com/api/projects/status/github/nhejazi/methyvim?branch=master&svg=true)](https://ci.appveyor.com/project/nhejazi/methyvim)
[![Coverage Status](https://img.shields.io/codecov/c/github/nhejazi/methyvim/master.svg)](https://codecov.io/github/nhejazi/methyvim?branch=master)
[![Project Status: Active â€“ The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![BioC status](http://www.bioconductor.org/shields/build/release/bioc/methyvim.svg)](https://bioconductor.org/checkResults/release/bioc-LATEST/methyvim)
[![Bioc Time](http://bioconductor.org/shields/years-in-bioc/methyvim.svg)](https://bioconductor.org/packages/release/bioc/html/methyvim.html)
[![Bioc Downloads](http://bioconductor.org/shields/downloads/methyvim.svg)](https://bioconductor.org/packages/release/bioc/html/methyvim.html)
[![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1401298.svg)](https://doi.org/10.5281/zenodo.1401298)

> Targeted, Robust, and Model-free Differential Methylation Analysis

__Authors:__ [Nima Hejazi](https://nimahejazi.org) and [Mark van der
  Laan](https://vanderlaan-lab.org/about)

---

## What's `methyvim`?

`methyvim` is an R package that provides facilities for differential methylation
analysis based on _variable importance measures_ (VIMs), statistical target
parameters inspired by causal inference.

The statistical methodology implemented computes targeted minimum loss estimates
of several well-characterized variable importance measures:

For discrete-valued treatments or exposures:

* The _average treatment effect_ (ATE): The effect of a binary exposure or
  treatment on the observed methylation at a target CpG site is estimated,
  controlling for the observed methylation at all other CpG sites in the same
  neighborhood as the target site, based on an additive form. In particular, the
  parameter estimate represents the __additive difference__ in methylation that
  would have been observed at the target site had all observations received the
  treatment versus the scenario in which none received the treatment.

* The _relative risk_ (RR): The effect of a binary exposure or treatment on the
  observed methylation at a target CpG site is estimated, controlling for the
  observed methylation at all other CpG sites in the same neighborhood as the
  target site, based on an geometric form. In particular, the parameter estimate
  represents the __multiplicative difference__ in methylation that would have
  been observed at the target site had all observations received the treatment
  versus the scenario in which none received the treatment.

For continuous-valued treatments or exposures (_WIP: support planned_):

* A _nonparametric variable importance measure_ (NPVI) [@chambaz2012estimation]:
  The effect of continous-valued exposure or treatment (the observed methylation
  at a target CpG site) on an outcome of interest is estimated, controlling for
  the observed methylation at all other CpG sites in the same neighborhood as
  the target (treatment) site, based on a parameter that compares values of the
  treatment against a reference value taken to be the null. In particular, the
  implementation provided is designed to assess the effect of differential
  methylation at the target CpG site on a (typically) phenotype-level outcome
  of interest (e.g., survival), in effect providing an nonparametric evaluation
  of the impact of methylation at the target site on said outcome.

These methods allow differential methylation effects to be quantified in a
manner that is largely assumption-free, especially of the variety exploited
in parametric models. __The statistical algorithm consists in several major
steps:__

1. Pre-screening of genomic sites is used to isolate a subset of sites for
  which there is cursory evidence of differential methylation. For the sake of
  computational feasibility, targeted minimum loss-based estimates of VIMs are
  computed only for this subset of sites. Several screening approaches are
  available, adapting core routines from the following R packages:
  [`limma`](http://bioconductor.org/packages/release/bioc/html/limma.html),
  [`tmle.npvi`](https://CRAN.R-project.org/package=tmle.npvi).
2. Nonparametric VIMs are estimated for the specified parameter, currently
  adapting routines from the
  [`tmle.npvi`](https://CRAN.R-project.org/package=tmle.npvi) and
  [`tmle`](https://CRAN.R-project.org/package=tmle) R packages.
3. Since pre-screening is performed prior to estimating VIMs, we make use of a
  multiple testing correction uniquely suited to such settings. Due to the
  multiple testing nature of the estimation problem, a variant of the Benjamini
  & Hochberg procedure for controlling the False Discovery Rate (FDR) is applied
  [@benjamini1995controlling]. Specifically, we apply the modified marginal
  Benjamini & Hochberg step-up False Discovery Rate controlling procedure for
  multi-stage analyses (FDR-MSA) [@tuglus2009modified].

For a general discussion of the framework of targeted minimum loss estimation
and its myriad applications, the canonical references are @vdl2011targeted and
@vdl2018targeted. @hernan2018causal and @pearl2000causality may be of interest
to those desiring a more general introduction to statistical causal inference.

---

## Installation

For standard use, install from
[Bioconductor](https://bioconductor.org/packages/methyvim) using
[`BiocManager`](https://CRAN.R-project.org/package=BiocManager):

```{r bioc-installation, eval = FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE)) {
  install.packages("BiocManager")
}
BiocManager::install("methyvim")
```

To contribute, install the bleeding-edge _development version_ from GitHub via
[`remotes`](https://CRAN.R-project.org/package=remotes):

```{r gh-master-installation, eval = FALSE}
remotes::install_github("nhejazi/methyvim")
```

Current and prior [Bioconductor](https://bioconductor.org) releases are
available under branches with numbers prefixed by "RELEASE_". For example, to
install the version of this package available via Bioconductor 3.6, use

```{r gh-develop-installation, eval = FALSE}
remotes::install_github("nhejazi/methyvim", ref = "RELEASE_3_6")
```

---

## Example

For details on how to best use the `methyvim` R package, please consult the most
recent [package
vignette](http://bioconductor.org/packages/release/bioc/vignettes/methyvim/inst/doc/using_methyvim.html)
available through the [Bioconductor
project](https://bioconductor.org/packages/methvim).

---

## Issues

If you encounter any bugs or have any specific feature requests, please [file an
issue](https://github.com/nhejazi/methyvim/issues).

---

## Contributions

Contributions are very welcome. Interested contributors should consult our
[contribution
guidelines](https://github.com/nhejazi/methyvim/blob/master/CONTRIBUTING.md)
prior to submitting a pull request.

---

## Citation

After using the `methyvim` R package, please cite the following:

        @article{hejazi2018methyvim,
          doi = {10.12688/f1000research.16047.1},
          url = {https://dx.doi.org/10.12688/f1000research.16047.1},
          year = {2018},
          publisher = {Faculty of 1000 Ltd},
          volume = {7},
          number = {1424},
          author = {Hejazi, Nima S and Phillips, Rachael V and Hubbard, Alan E
            and {van der Laan}, Mark J},
          title = {{methyvim}: Targeted, robust, and model-free differential
            methylation analysis in {R}},
          journal = {F1000Research}
        }

        @manual{hejazi2019methyvimbioc,
          author = {Hejazi, Nima S and {van der Laan}, Mark J},
          title = {{methyvim}: Targeted, robust, and model-free differential
            methylation analysis},
          doi = {10.18129/B9.bioc.methyvim},
          url = {https://bioconductor.org/packages/methyvim},
          note = {R package version 1.6.0}
        }

---

## Related

* [R/`methyvimData`](https://github.com/nhejazi/methyvimData) - R package with
    sample experimental DNA methylation data for use as an example with this
    analysis package.

---

## Funding

The development of this software was supported in part through grants from the
National Institutes of Health: [T32 LM012417-02](https://projectreporter.nih.gov/project_info_description.cfm?aid=9248418&icde=37849831&ddparam=&ddvalue=&ddsub=&cr=1&csb=default&cs=ASC&pball=), [R01 ES021369-05](https://projectreporter.nih.gov/project_info_description.cfm?aid=9210551&icde=37849782&ddparam=&ddvalue=&ddsub=&cr=1&csb=default&cs=ASC&pball=), and [P42 ES004705-29](https://projectreporter.nih.gov/project_info_details.cfm?aid=9260357&map=y).

---

## License

&copy; 2017-2019 [Nima S. Hejazi](https://nimahejazi.org)

The contents of this repository are distributed under the MIT license. See file
`LICENSE` for details.

---

## References

