---
title: "*F1000Research* Software Tool Article Template"
author:
  - name: Nima S. Hejazi
    affiliation: Address of author1
  - name: Rachael Phillips
    affiliation: Address of author1
  - name: Alan E. Hubbard
    affiliation: Address of author2
  - name: Mark J. van der Laan
    affiliation: Address of author2
abstract: Abstracts should be up to 300 words and provide a succinct summary of
  the article. Although the abstract should explain why the article might be
  interesting, care should be taken not to inappropriately over-emphasise the
  importance of the work described in the article. Citations should not be used
  in the abstract, and the use of abbreviations should be minimized.
keywords: DNA methylation, differential methylation, epigenetics, causal
  inference, variable importance, targeted minimum loss-based estimation.
bibliography: paper_BiocF1000.bib
output: BiocWorkflowTools::f1000_article
---

# Introduction

The introduction provides context as to why the software tool was developed and
what need it addresses. It is good scholarly practice to mention previously
developed tools that address similar needs, and why the current tool is needed.

# Methods
## Implementation
For software tool papers, this section should address how the tool works and any
relevant technical details required for implementation of the tool by other
developers.

## Operation
This part of the methods should include the minimal system requirements needed
to run the software and an overview of the workflow for the tool for users of
the tool.


# Results <!-- Optional - only if novel data or analyses are included -->
This section is only required if the paper includes novel data or analyses, and
should be written as a traditional results section.

# Use Cases <!-- Optional - only if NO new datasets are included -->
This section is required if the paper does not include novel data or analyses.
Examples of input and output files should be provided with some explanatory
context.  Any novel or complex variable parameters should also be explained in
sufficient detail to allow users to understand and use the tool's functionality.

# Discussion <!-- Optional - only if novel data or analyses are included -->
This section is only required if the paper includes novel data or analyses, and
should be written in the same style as a traditional discussion section.
Please include a brief discussion of allowances made (if any) for controlling
bias or unwanted sources of variability, and the limitations of any novel
datasets.


# Conclusions <!-- Optional - only if novel data or analyses are included -->
This section is only required if the paper includes novel data or analyses, and
should be written as a traditional conclusion.

# Summary <!-- Optional - only if NO new datasets are included -->
This section is required if the paper does not include novel data or analyses.
It allows authors to briefly summarize the key points from the article.

# Data availability <!-- Optional - only if novel data or analyses are included -->
Please add details of where any datasets that are mentioned in the paper, and
that have not have not previously been formally published, can be found. If
previously published datasets are mentioned, these should be cited in the
references, as per usual scholarly conventions.

# Software availability
This section will be generated by the Editorial Office before publication.
Authors are asked to provide some initial information to assist the Editorial
Office, as detailed below.

1. URL link to where the software can be downloaded from or used by a non-coder (AUTHOR TO PROVIDE; optional)
2. URL link to the author's version control system repository containing the source code (AUTHOR TO PROVIDE; required)
3. Link to source code as at time of publication (*F1000Research* TO GENERATE)
4. Link to archived source code as at time of publication (*F1000Research* TO GENERATE)
5. Software license (AUTHOR TO PROVIDE; required)


# Author contributions
In order to give appropriate credit to each author of an article, the individual
contributions of each author to the manuscript should be detailed in this
section. We recommend using author initials and then stating briefly how they
contributed.

# Competing interests
All financial, personal, or professional competing interests for any of the
authors that could be construed to unduly influence the content of the article
must be disclosed and will be displayed alongside the article. If there are no
relevant competing interests to declare, please add the following: 'No competing
interests were disclosed'.

# Grant information
Please state who funded the work discussed in this article, whether it is your
employer, a grant funder etc. Please do not list funding that you have that is
not relevant to this specific piece of research. For each funder, please state
the funder’s name, the grant number where applicable, and the individual to whom
the grant was assigned. If your work was not funded by any grants, please
include the line: 'The author(s) declared that no grants were involved in
supporting this work.'

# Acknowledgments
This section should acknowledge anyone who contributed to the research or the
article but who does not qualify as an author based on the criteria provided
earlier (e.g. someone or an organization that provided writing assistance).
Please state how they contributed; authors should obtain permission to
acknowledge from all those mentioned in the Acknowledgments section.

Please do not list grant funding in this section.

# USING R MARKDOWN
Some examples of commonly used markdown syntax are listed below, to help you get
started.

## Cross-references
For portability between different output formats, use the syntax introduced by
*bookdown*, such as `(\#label)` for labels and `\@ref(label)` for
cross-references. The following sections provide examples of referencing tables,
figures, and equations.

## Citations

You can include refences in a standard Bibtex file.  The name of this file is
given in the header of the markdown document (in our case it is *sample.bib*).
References to entries in the Bibtex file are made using square brackets and use
an @ plus the key for the entry you are referencing [@Smith:2012qr].  You can
combine multiple entries by separating them with a semi-colon
[@Smith:2012qr; @Smith:2013jd].

## Code chunks
You can embed an R code chunk like this:
```{r firstCodeChunk}
x <- 1:10
x
```

If you specify a figure caption to a code chunk using the chunk option `fig.cap`, the plot will be automatically labeled and numbered. The figure label is generated from the label of the code chunk by prefixing it with `fig:`, e.g., see Figure \@ref(fig:plot).

```{r plot, fig.width=4, fig.height=4, fig.cap='A caption to our sample plot.'}
plot(x)
```

## Tables

Markdown syntax tends to lack some of the more sophisticated formatting features available in LaTeX, so you may need to edit the tables later to get the desired format.

| First name  | Last Name | Grade |
| ----------- | --------- | ----- |
| John        | Doe       |   7.5 |
| Richard     | Miles     |     2 |

Table: Caption to table.

Just like figures, tables with captions will also be numbered and can be referenced. Captions are entered as a paragraph beginning with the string "Table:" (or just ":"), which may appear either before or after the table. A label for the table should appear in the beginning of the caption in the form of `(\#tab:label)`, e.g., see Table \@ref(tab:table).

: (\#tab:table) A table with text justification.

| First name  | Last Name | Grade |
| ----------- | :-------: | ----: |
| John        | Doe       |   7.5 |
| Richard     | Miles     |     2 |

## Figures

You can include static figures (i.e. no generated by code) using the `include_graphics()` function from the **knitr** package, in a standard code chunk.

```{r frog-picture, out.width='50%', fig.align='center', fig.cap = 'Your figure legend goes here; it should be succinct, while still explaining all symbols and abbreviations.'}
knitr::include_graphics('frog.jpg')
```

You can again use the `fig.cap` option to provide the figure caption, and reference the image based on the code chunk label.  You can also use options such as `fig.align` and `fig.width` to adjust the position and size of the image within the final document, e.g. Figure \@ref(fig:frog-picture) is a frog.

Alternatively, you can use the standard markdown syntax like so:

![This is a smaller version of the same picture, inserted using the standard markdown syntax](frog.jpg){width=25%}

Please give figures appropriate filenames, e.g.: figure1.pdf, figure2.png.

Figure legends should briefly describe the key messages of the figure such that the figure can stand alone from the main text. However, all figures should also be discussed in the article text. Each legend should have a concise title of no more than 15 words. The legend itself should be succinct, while still explaining all symbols and abbreviations. Avoid lengthy descriptions of methods.

For any figures reproduced from another publication (as long as appropriate permission has been obtained from the copyright holder —see under the heading 'Submission'), please include a line in the legend to state that: 'This figure has been reproduced with kind permission from [include original publication citation]'.

## Mathematics

You can use LaTeX syntax to typeset mathematical expressions.  Let $X_1, X_2, \ldots, X_n$ be a sequence of independent and identically distributed random variables with $\text{E}[X_i] = \mu$ and $\text{Var}[X_i] = \sigma^2 < \infty$, and let
$$S_n = \frac{X_1 + X_2 + \cdots + X_n}{n}
      = \frac{1}{n}\sum_{i}^{n} X_i$$
denote their mean. Then as $n$ approaches infinity, the random variables $\sqrt{n}(S_n - \mu)$ converge in distribution to a normal $\mathcal{N}(0, \sigma^2)$.

To number and refer to equations, put them in the equation environments and assign labels to them, as for Equation \@ref(eq:binom).

\begin{equation}
  f\left(k\right) = \binom{n}{k} p^k\left(1-p\right)^{n-k}
  (\#eq:binom)
\end{equation}

## Lists

You can make ordered lists

1. Like this,
2. and like this.

or bullet points

* Like this,
* and like this.
    + Use indentation
        - for sub-items



DNA methylation is a fundamental epigenetic process known to play an important
role in controlling gene expression. It remains perhaps the best studied
biological mechanism amongst all known epigenetic processes of its kind. With
relatively recent advances in computational biology and bioinformatics, probing
DNA methylation has become rather easily possible -- in fact, modern assays
allow for DNA methylation signatures at up to $850,000$ CpG sites to be measured
simultaneously. Most statistical techniques available for the analysis of data
produced by such assays rely on (generalized) linear models. Here, we provide an
alternative to such approaches. Specifically, we provide a range of _variable
importance measures_ (__VIM__), parameters that arise in statistical causal
inference, for which targeted minimum loss-based estimates may be readily
computed based on the data made available by DNA methylation assays. `methyvim`
is an R package that provides facilities for performing differential methylation
analyses within exactly this scope.

For a general discussion of the framework of targeted minimum loss-based
estimation and the role this approach plays in statistical causal inference, the
interested reader is invited to consult @vdl2011targeted and @vdl2017targeted.
For a more general introduction to (statistical) causal inference,
@pearl2009causality and @hernan2018causal may be of interest.

---

## Methodology

The core functionality of this package is made available via the eponymous
`methyvim` function, which implements a statistical algorithm designed to
compute targeted estimates of VIMs, defined in such a way that the VIMs
represent parameters of scientific interest in computational biology
experiments; moreover, these VIMs are defined such that they may be estimated in
a manner that is very nearly assumption-free, that is, within a fully
nonparametric statistical model. __The statistical algorithm consists in several
major steps:__

1. Pre-screening of genomic sites is used to isolate a subset of sites for
  which there is cursory evidence of differential methylation. For the sake of
  computational feasibility, targeted minimum loss-based estimates of VIMs are
  computed only for this subset of sites. Currently, the available screening
  approaches adapts core routines from the
  [`limma`](http://bioconductor.org/packages/release/bioc/html/limma.html) R
  package, thought future releases will support functionality from other
  packages (e.g., [`tmle.npvi`](https://CRAN.R-project.org/package=tmle.npvi)).

2. Nonparametric estimates of VIMs (for the specified target parameter) are
  currently computed by adapting routines from the
  [`tmle`](https://CRAN.R-project.org/package=tmle) R package. Future releases
  will support doubly-robust estimates of these VIMs (via the
  [`drtmle`](https://cran.r-project.org/web/packages/drtmle/index.html) package)
  and add parameters for continuous treatments/exposures (via the
  [`tmle.npvi`](https://CRAN.R-project.org/package=tmle.npvi) package).

3. Since pre-screening is performed prior to estimating VIMs, we make use of a
  multiple testing correction uniquely suited to such settings. Due to the
  multiple testing nature of the estimation problem, a variant of the Benjamini
  & Hochberg procedure for controlling the False Discovery Rate (FDR) is applied
  [@benjamini1995controlling]. Specifically, we apply the modified marginal
  Benjamini & Hochberg step-up False Discovery Rate controlling procedure for
  multi-stage analyses (FDR-MSA), which is guaranteed to control the FDR as if
  all sites were tested [@tuglus2009modified].

---

## Parameters of Interest

For discrete-valued treatments or exposures:

* The _average treatment effect_ (ATE): The effect of a binary exposure or
  treatment on the observed methylation at a target CpG site is estimated,
  controlling for the observed methylation at all other CpG sites in the same
  neighborhood as the target site, based on an additive form. In particular, the
  parameter estimate represents the __additive difference__ in methylation that
  would have been observed at the target site had all observations received the
  treatment versus the scenario in which none received the treatment.

* The _relative risk_ (RR): The effect of a binary exposure or treatment on the
  observed methylation at a target CpG site is estimated, controlling for the
  observed methylation at all other CpG sites in the same neighborhood as the
  target site, based on an geometric form. In particular, the parameter estimate
  represents the __multiplicative difference__ in methylation that would have
  been observed at the target site had all observations received the treatment
  versus the scenario in which none received the treatment.

Support for continuous-valued treatments or exposures is _planned but not yet
available_, though work is underway to incorporate into our methodology the
following

* A _nonparametric variable importance measure_ (NPVI) [@chambaz2012estimation]:
  The effect of continous-valued exposure or treatment (the observed methylation
  at a target CpG site) on an outcome of interest is estimated, controlling for
  the observed methylation at all other CpG sites in the same neighborhood as
  the target (treatment) site, based on a parameter that compares values of the
  treatment against a reference value taken to be the null. In particular, the
  implementation provided is designed to assess the effect of differential
  methylation at the target CpG site on a (typically) phenotype-level outcome
  of interest (e.g., survival), in effect providing an nonparametric evaluation
  of the impact of methylation at the target site on said outcome.

_As previously noted, in all cases, an estimator of the target parameter is
constructed via targeted minimum loss-based estimation._

Having now discussed the foundational principles of the estimation procedure
employed and the statistical algorithm implemented, it is best to proceed by
examining `methyvim` by example.

---

## Preliminaries: Setting up the Data

```{r reqs, echo=FALSE}
suppressMessages(library(tmle))
suppressMessages(library(minfi))
suppressMessages(library(SummarizedExperiment))
```

First, we'll load the `methyvim` package and the example data contained in the
`methyvimData` package that accompanies it:

```{r prelims}
library(methyvim)
library(methyvimData)
```

Now, let's load the data set and seed the RNG:

```{r get-data}
data(grsExample)
set.seed(479253)
grsExample
```

The example data object is of class `GenomicRatioSet`, provided by the `minfi`
package. The summary provided by the `print` method gives a wealth of
information on the experiment that generated the data -- since we are working
with a simulated data set, we need not concern ourselves with much of this
information. While most slots of objects of this S4 class have descriptive
names, the interested analyst might consider consulting the documentation of the
`minfi` package for an in-depth description [@aryee2014minfi].

---

## Quantifying the Effect of an Exposure on DNA Methylation

### The Average Treatment Effect as Variable Importance Measure

...

```{r methyvim-ate}
suppressMessages(
  methyvim_out_ate <- methyvim(data_grs = grsExample, sites_comp = 10,
                               var_int = 1, vim = "ate", type = "Mval",
                               filter = "limma", filter_cutoff = 0.05,
                               parallel = FALSE, tmle_type = "sl"
                              )
)
methyvim_out_ate
```

As is clear from examining the object `methyvim_out_ate`, the output resembles
exactly that returned when examining objects of class `GenomicRatioSet` from the
`minfi` R package. In particular, the returned `methytmle` object is merely a
modified form (in particular, a subclass) of the input `GenomicRatioSet` object
-- thus, it contains all of the original slots, with all experimental data
intact. Several extra pieces of information are contained within the output
object as well_.

We can take a look at the results produced from the estimation procedure by
examining the `"vim"` `slot` of the `methytmle` object:

```{r methyvim-ate-print}
head(slot(methyvim_out_ate, "vim"))
```

From the table displayed, we note that we have access to point estimates of the
ATE ("est_ATE") as well as lower and upper confidence interval bounds for each
estimate ("lowerCI_ATE" and "upperCI_ATE", respectively). Additional statistical
information we have access to include the variance ("Var_ATE") of the estimate
as well as the p-value ("pval") associated with each estimate (based on
Wald-style testing procedures). Beyond these, key bioinformatical quantities
(with respect to the algorithm outlined above) are also returned; these include
the total number of neighbors of the target site, the number of neighboring
sites controlled for when estimating the effect of exposure on DNA methylation,
and, finally, the maximum correlation between the target site and any given site
in its full set of neighbors.

...

---

### The Risk Ratio as Variable Importance Measure

...

```{r methyvim-rr}
methyvim_out_rr <- methyvim(data_grs = grsExample, sites_comp = 10,
                            var_int = 1, vim = "rr", type = "Mval",
                            filter = "limma", filter_cutoff = 0.05,
                            parallel = FALSE, tmle_type = "sl"
                           )
methyvim_out_rr
```
...

```{r methyvim-rr-print}
head(slot(methyvim_out_rr, "vim"))
```
...

---

## Visualization of Results

...

```{r methyvim-plot-volcano, eval=FALSE}
plot(methyvim_out_rr, type = "volcano")
```

...

```{r methyvim-plot-pvals, eval=FALSE}
plot(methyvim_out_rr, type = "pvals")
```

...

```{r methyvim-plot-heatmap, eval=FALSE}
plot(methyvim_out_rr, type = "heatmap")
```

...

---

## References


